spacy/ml/models/entity_linker.py

```diff
@@ -78,23 +80,36 @@ def span_maker_forward(model, docs: List[Doc], is_train) -> Tuple[Ragged, Callab
 
 
 @registry.misc("spacy.KBFromFile.v1")
-def load_kb(kb_path: Path) -> Callable[[Vocab], KnowledgeBase]:
-    def kb_from_file(vocab):
-        kb = KnowledgeBase(vocab, entity_vector_length=1)
+def load_kb(
+    kb_path: Path,
+) -> Callable[[Vocab], InMemoryLookupKB]:
+    def kb_from_file(vocab: Vocab):
+        kb = InMemoryLookupKB(vocab, entity_vector_length=1)
         kb.from_disk(kb_path)
         return kb
 
     return kb_from_file
 
 
 @registry.misc("spacy.EmptyKB.v1")
-def empty_kb(entity_vector_length: int) -> Callable[[Vocab], KnowledgeBase]:
-    def empty_kb_factory(vocab):
-        return KnowledgeBase(vocab=vocab, entity_vector_length=entity_vector_length)
+def empty_kb(
+    entity_vector_length: int,
+) -> Callable[[Vocab], InMemoryLookupKB]:
+    def empty_kb_factory(vocab: Vocab):
+        return InMemoryLookupKB(vocab=vocab, entity_vector_length=entity_vector_length)
 
     return empty_kb_factory
 
 
 @registry.misc("spacy.CandidateGenerator.v1")
-def create_candidates() -> Callable[[KnowledgeBase, Span], Iterable[Candidate]]:
+def create_candidates() -> Callable[
+    [KnowledgeBase, Union[Span, str]], Iterator[Candidate]
+]:
```

Comment:
Why do we need to change the types for this one?

Reply:
If we really need a change of types, I'd prefer a version bump.

Reply:
We don't _need_ one, but the `KnowledgeBase` interface currently supports `Union[Span, str]` and I considered this a non-breaking change (guess type checks might fail though?). 
Should we go with `Span` instead? The motivation was that it might be more convenient to pass only strings instead of `Span`s and that it should be up to the `KnowledgeBase` implementation. I don't mind requiring `Span`s though either.

Reply:
(there's not just the `Span` but also the `Iterator` change)

Reply:
`Iterator` change has been reverted.

Reply:
I think requiring `Span`'s is fine - you'd expect there is always some text-bound mention which has offsets, making it a `Span` and defining the context in case the candidate generator wants to use it.